### ISSUE TITLE
Introduce `number_format_i18n` for @wordpress/i18n

### ISSUE THREAD

**Issue by Clorith** [WordPress/gutenberg#22628]
WordPress translation functions have had [`number_format_i18n`](https://developer.wordpress.org/reference/functions/number_format_i18n/) for some time, which is convenient for outputting numbers in a locale-appropriate way.

It would be beneficial to discuss the possibilities of introducing this as part of the `@wordpress/i18n` package as well.

**Comment by swissspidy** [comment]
This could be a small wrapper around [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat) that handles the locale part.

**Comment by pedro-mendonca** [comment]
Hi, any development on this?

**Comment by USERSATOSHI** [comment]
Hi, From my understanding of this issue, I believe we can use 

`getSettings()` from `@wordpress/date` which would provide us with the locale.
This locale can be converted into JS Compatible format and can be used internally in `numberFormatI18n` function 

allowing us to have a similar code as PHP.

an example Implementation for this could be:

```ts
export function numberFormatI18n(
	number: number,
	decimals: number = 0
): string {
	const locale = getLocale(); // An internal helper function to get locale.

	/**
	 * Node.js supports up to 20 decimal places for Intl.NumberFormat ( Tested manually )
	 * but most browsers support up to 100 decimal places. ( Mdn Docs)
	 * We will use 20 as a safe upper limit for compatibility.
	 */
	const validDecimals = Math.max( 0, Math.min( 20, Math.floor( decimals ) ) );

	const options: Intl.NumberFormatOptions = {
		minimumFractionDigits: validDecimals,
		maximumFractionDigits: validDecimals,
	};

	return new Intl.NumberFormat( locale, options ).format( number );
}
```

if this implementation is fine, I can go ahead and create a PR for this.

cc: @swissspidy.

**Comment by swissspidy** [comment]
It's not that trivial. 

First, we shouldn't add a dependency on the date package to get locale information. This is the i18n package after all, which itself should have locale information. That's the whole purpose. So we need to figure out how to do that. Maybe we even need a change in core?

Second, the WordPress locale would need to be converted to a BCP 47 language tag or `Intl.Locale` instance supported by `Intl.NumberFormat`. That's easy for e.g. `de_DE` -> `de-DE` but there are also variants such as `de_DE_formal` and some other [WordPress/GlotPress-specific variants](https://github.com/GlotPress/GlotPress/blob/8f0f877107db7a5d41bc0ebfc11a5b24ad8e2f8d/locales/locales.php) that need to be supported and tested.

**Comment by USERSATOSHI** [comment]
> First, we shouldn't add a dependency on the date package to get locale information. This is the i18n package after all, which itself should have locale information. That's the whole purpose. So we need to figure out how to do that. Maybe we even need a change in core?

Oh , yeah if we are not depending on date then this wouldn't be trivial , I saw locale info in that hence thought it could be a nice way to add it, I will check on this one and see what we can do to get locale without any deps

> Second, the WordPress locale would need to be converted to a BCP 47 language tag or Intl.Locale instance supported by Intl.NumberFormat. That's easy for e.g. de_DE -> de-DE but there are also variants such as de_DE_formal and some other [WordPress/GlotPress-specific variants](https://github.com/GlotPress/GlotPress/blob/8f0f877107db7a5d41bc0ebfc11a5b24ad8e2f8d/locales/locales.php) that need to be supported and tested.

for this I was able to create a wpLocaleToJS function which doesnt convert it tho I have to check the variants ( thank you for sharing that ) to test on all locale and support anything missing

Right now it supports the lang[_Script][Country] format.

**Comment by USERSATOSHI** [comment]
One thing we could possible do, is to add numberFormatI18n to `createI18n` and `defaultI18n` as these create their own instances and 

```ts
export const createI18n = < TextDomain extends string >(
	initialData?: LocaleData< TextDomain >,
	initialDomain?: TextDomain,
	hooks?: Hooks
): I18n< TextDomain >
```
does take LocaleData 

which include an optional lang: property in it.

Link to type: https://github.com/aduth/tannin/blob/master/packages/tannin/index.js#L13-L19

**Comment by swissspidy** [comment]
Right, yeah, in WP we usually provide `lang`, which is iso_639 + country code. Looks like it doesn't include the variant suffix like `_formal`, so even better.

**Comment by USERSATOSHI** [comment]
> Right, yeah, in WP we usually provide `lang`, which is iso_639 + country code. Looks like it doesn't include the variant suffix like `_formal`, so even better.

Yeah, I just wanted to confirm how locale was being set in WordPress.

<img width="472" height="308" alt="Image" src="https://github.com/user-attachments/assets/350dbe41-5ebe-442b-9193-e252425bfa26" />

Seems like the we can add numberFormatI18n as part of createI18n and defaultI18n.

cc: @swissspidy

**Comment by USERSATOSHI** [comment]
In the mean time, I tried to implement this in createI18n and defaultI18n

```ts
	const numberFormatI18n: I18n[ 'numberFormatI18n' ] = (
		number,
		decimals = 0,
		domain = 'default' as TextDomain
	) => {
		const localeData = ( getLocaleData( domain )?.[
			''
		] as I18nDomainMetadata< TextDomain > ) ?? {
			lang: 'en_US',
			domain: 'default',
			plural_forms: 'n === 1 ? 0 : 1',
		};

		let wpLocale = localeData.lang || 'en_US';
		let jsLocale;
		if ( ! I18N_WP_LOCALE_REGEXP.test( wpLocale ) ) {
			jsLocale = 'en-US';
		} else {
			jsLocale = wpLocale.replace( '_', '-' );
		}
		const validDecimals = Math.max( 0, Math.min( decimals, 20 ) );

		const options: Intl.NumberFormatOptions = {
			minimumFractionDigits: validDecimals,
			maximumFractionDigits: validDecimals,
		};

		return new Intl.NumberFormat( jsLocale, options ).format( number );
	};
```

I think, we can take this approach for numberFormatI18n as this matches our criteria for no deps for locale.

UPDATE: slightly updated the function to accept domains in createI18n , defaultI18n doesn't have domain field. This was done due to support for custom domain in createI18n.

**Comment by USERSATOSHI** [comment]
Hi @swissspidy, I have created a prototype implementation on basis of the discussion and would love an opinion If what I understood from this is correct or not as well as if there anything we also might need to consider.

**Comment by dmytropetrash** [comment]
Hi! Thanks for opening this issue. To introduce `number_format_i18n` in `@wordpress/i18n`, a good next step would be to propose a PR adding the function with necessary tests. Let me know if you need any help getting started!
